designs/operator-monitoring-plugin.md

```diff
@@ -0,0 +1,209 @@
+---
+title: controller-observation-plugin
+authors:
+  - "@Kavinjsir"
+reviewers:
+  - "@varsha"
+  - "@camila"
+approvers:
+creation-date: 2022-06-06
+last-updated: 2022-06-06
+status: provisional
+see-also:
+  - ""
+replaces:
+  - ""
+superseded-by:
+  - ""
+---
+
+# a-grafana-dashboard-plugin-to-observe-controller-metrics
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Open Questions [optional]
+
+- TBD
+
+## Summary
+
+This EP will provide a way for the operator author to visualize the runtime metrics in order to observe the operator status.
+
+The functionality will be implemented by a new plugin that can generate the 'manifests' to be loaded on Grafana.
+
+In the context of this EP, the Grafana 'manifests' are mainly:
+
+- JSON files to be directly loaded on Grafana dashboards
+- TBD
+
+## Motivation
+
+The [controller-runtime](https://github.com/kubernetes-sigs/controller-runtime) provides rich metrics to measure controllers' status from various perspectives.
+
+Currently, users of Kubebuilder need to design and create their own Grafana dashboards to visualize the metrics.
+It would be great if we can provide a unified and comprehensive approach so that users may:
+
+- easily visualize controller/operator metrics
+- avoid duplicate creation of Grafana dashboards file
+- improve the observability towards the opeartor, keep it safe and robust
+
+### Goals
+
+- Provide manifests that can display operator status for common needs.
+- Add a plugin or enhance the current kustomize plugin that scaffolds Grafana manifests.
+
+### Non-Goals
+
+- TBD
```

Comment:
Some non-goals could be:
1. We do not deal with the kind of metrics which will be exported.
2. The proposal will talk not only about scaffolding basic manifests (or templates) for integrating grafana dashboards. It does not deal with dashboard designs.